pkg/store/proxy_merge.go

```diff
@@ -181,29 +191,52 @@ func NewProxyResponseLoserTree(seriesSets ...respSet) *losertree.Tree[*storepb.S
 		if a == maxVal && b == maxVal {
 			return true
 		}
-		if a.GetSeries() != nil && b.GetSeries() != nil {
-			iLbls := labelpb.ZLabelsToPromLabels(a.GetSeries().Labels)
-			jLbls := labelpb.ZLabelsToPromLabels(b.GetSeries().Labels)
+		ar, br := a.response, b.response
 
-			return labels.Compare(iLbls, jLbls) < 0
-		} else if a.GetSeries() == nil && b.GetSeries() != nil {
+		if ar.GetSeries() != nil && br.GetSeries() != nil {
+			aLbls := labelpb.ZLabelsToPromLabels(ar.GetSeries().Labels)
+			bLbls := labelpb.ZLabelsToPromLabels(br.GetSeries().Labels)
+
+			if cmp := labels.Compare(aLbls, bLbls); cmp == 0 {
+				// use store labelset as tiebreaker for a stable order of responses and predictable deduplication
+				return strings.Compare(a.labelset, b.labelset) < 0
+			} else {
+				return cmp < 0
+			}
+		} else if ar.GetSeries() == nil && br.GetSeries() != nil {
 			return true
-		} else if a.GetSeries() != nil && b.GetSeries() == nil {
+		} else if ar.GetSeries() != nil && br.GetSeries() == nil {
 			return false
 		}
 
-		if a.GetWarning() != "" && b.GetWarning() != "" {
-			return len(a.GetWarning()) < len(b.GetWarning())
+		if ar.GetWarning() != "" && br.GetWarning() != "" {
+			return len(ar.GetWarning()) < len(br.GetWarning())
 		}
-
 		return false
 	}
 
-	return losertree.New[*storepb.SeriesResponse, respSet](seriesSets, maxVal, func(s respSet) *storepb.SeriesResponse {
-		return s.At()
-	}, less, func(s respSet) {
-		s.Close()
-	})
+	return &proxyResponseLoserTree{
+		tree: losertree.New(
+			seriesSets,
+			maxVal,
+			func(s respSet) seriesResponseWithStoreLabelset {
```

Comment:
At() is called more than once for every item in the set. Could we upgrade `type respSet interface` to be:

```
type respSet interface {
	Close()
	At() seriesResponseWithStoreLabelset
```

?

Plus, `s.Labelset()` allocates a new string each time. That's not good in a hot path. Internally, it's just:

```
storeLabelSets []labels.Labels
```

I think it should be much faster to compare these things directly instead of allocating a bunch of strings that will be identical across millions of series +/-.